### ISSUE TITLE
RuntimeError: svd_cuda: the updating process of SBDSDC did not converge (error: 1)

### ISSUE THREAD

**Issue by ryh95** [pytorch/pytorch#28293]
Hi,

when I run torch.svd() on a matrix with GPU, it raises the error
`RuntimeError: svd_cuda: the updating process of SBDSDC did not converge (error: 1)`

However, the torch.svd() has no problem with the matrix on CPU

Can someone help me figure out the reason?
Thank you!

The matrix is attached
[runtime_error_W.zip](https://github.com/pytorch/pytorch/files/3744970/runtime_error_W.zip)

**Comment by vishwakftw** [comment]
Hi @ryh95, thank you for opening the issue.

This seems to an issue with MAGMA, possibly relating to the correctness of `gesdd` vs `gesvd`. Note: the matrix is well-conditioned, but has too many repeated singular values.

**Comment by ryh95** [comment]
It seems that `gesdd` is used for both CPU and GPU versions of torch.svd(), according to the documentation.

> The implementation of SVD on CPU uses the LAPACK routine ?gesdd (a divide-and-conquer algorithm) instead of ?gesvd for speed. Analogously, the SVD on GPU uses the MAGMA routine gesdd as well.

So, why this relate to the correctness of `gesdd` vs `gesvd`?

**Comment by Hippogriff** [comment]
Is there a temporary solution/work-around to this problem?

**Comment by ryh95** [comment]
I just move the tensor to CPU then use `torch.svd` to get a decomposition. I'll move the results back to GPU if needed.

Someone has mentioned that `gesvd` is more accurate/robust(https://github.com/pytorch/pytorch/issues/25978#issue-492018796), so you can move the tensor to CPU and transfer to numpy array and use scipy if you prefer a better result.

**Comment by Hippogriff** [comment]
@ryh95 I also want to back propagate the gradients through this computation. Converting it to cpu might break the computation graph (though I am not sure about this).

**Comment by Nav94** [comment]
@Hippogriff Did you manage to work around the issue? I am also running into the same issue and was thinking of computing on the cpu and moving it back to the gpu.

**Comment by Hippogriff** [comment]
@Nav94 This can happens either matrix is severely ill conditioned or because the singular values are very close or equal to each other.
There are following solution to the problem:
1. For ill conditioned case, you can compute the condition number of the matrix on cpu and if the condition number is very large, then you cannot do much. In this case, you can simply trivialize the solution.
2. If the singular values are close to each other, you need to safe guard your back prop, that is, you need to write a new back ward pass. You can use the custom_svd function that replaces the torch's svd function.

`

    def compute_grad_V(U, S, V, grad_V):
        N = S.shape[0]
        K = svd_grad_K(S)
        S = torch.eye(N).cuda(S.get_device()) * S.reshape((N, 1))
        inner = K.T * (V.T @ grad_V)
        inner = (inner + inner.T) / 2.0
        return 2 * U @ S @ inner @ V.T


    def svd_grad_K(S):
        N = S.shape[0]
        s1 = S.view((1, N))
        s2 = S.view((N, 1))
        diff = s2 - s1
        plus = s2 + s1

        # TODO Look into it
        eps = torch.ones((N, N)) * 10**(-6)
        eps = eps.cuda(S.get_device())
        max_diff = torch.max(torch.abs(diff), eps)
        sign_diff = torch.sign(diff)

        K_neg = sign_diff * max_diff

        # gaurd the matrix inversion
        K_neg[torch.arange(N), torch.arange(N)] = 10 ** (-6)
        K_neg = 1 / K_neg
        K_pos = 1 / plus

        ones = torch.ones((N, N)).cuda(S.get_device())
        rm_diag = ones - torch.eye(N).cuda(S.get_device())
        K = K_neg * K_pos * rm_diag
        return K


    class CustomSVD(Function):
        """
        Costum SVD to deal with the situations when the
        singular values are equal. In this case, if dealt
        normally the gradient w.r.t to the input goes to inf.
        To deal with this situation, we replace the entries of
        a K matrix from eq: 13 in https://arxiv.org/pdf/1509.07838.pdf
        to high value.
        Note: only applicable for the tall and square matrix and doesn't
        give correct gradients for fat matrix. Maybe transpose of the
        original matrix is requires to deal with this situation. Left for
        future work.
        """
        @staticmethod
        def forward(ctx, input):
            # Note: input is matrix of size m x n with m >= n.
            # Note: if above assumption is voilated, the gradients
            # will be wrong.
            try:
                U, S, V = torch.svd(input, some=True)
            except:
                import ipdb; ipdb.set_trace()

            ctx.save_for_backward(U, S, V)
            return U, S, V

        @staticmethod
        def backward(ctx, grad_U, grad_S, grad_V):
            U, S, V = ctx.saved_tensors
            grad_input = compute_grad_V(U, S, V, grad_V)
            return grad_input

    customsvd = CustomSVD.apply
`

**Comment by ryh95** [comment]
Thanks for your solution! @Hippogriff 
By the way, is it true that convert the tensor to CPU will break the backpropagate?

**Comment by Hippogriff** [comment]
Yes, as far as I know.

**Comment by Abdelpakey** [comment]
@Hippogriff 
Thanks for the solution, is there any way to track your code in case you update it for the #TODO part?

**Comment by Hippogriff** [comment]
@Abdelpakey I am not keeping track of this yet. There are severl todos, for example, this code doesn't compute the gradient with respect to right and left eigen vectors. I will update this repo when I am done:
https://github.com/Hippogriff/SVD-Pytorch

**Comment by chenhao1umbc** [comment]
I am using this, since it is not solved
```
    try:
        u, s, v = torch.svd(L)
    except:                     ########## and torch.svd may have convergence issues for GPU and CPU.
        u, s, v = torch.svd(L + 1e-4*L.mean()*torch.rand(l, h))
```

**Comment by SebastianGrans** [comment]
@chenhao1umbc Thanks for that snippet! 

Is there a particular reason why you multiply with `L.mean()`?

**Comment by chenhao1umbc** [comment]
Yes, the main idea is that convergence issue can be solved by adding some turbulence. But the scale of L is unknown, which means L could be at 1e4 scale or 1e-4 scale. This means we cannot simple adding a "small" random number, but a relatively small number.

**Comment by wangyujiewj** [comment]
> I am using this, since it is not solved
> 
> ```
>     try:
>         u, s, v = torch.svd(L)
>     except:                     # torch.svd may have convergence issues for GPU and CPU.
>         u, s, v = torch.svd(L + 1e-4*L.mean()*torch.rand(l, h))
> ```

So how to set the values of l and h?

**Comment by davidbau** [comment]
This affects torch.pinverse also, due to its underlying svd.

**Comment by andreaskoepf** [comment]
For pytorch 1.8.1+cu101 the output of:
```
x = torch.randn(64, 64).cuda()
x[0,0] = float('nan')
torch.svd(x)
```
`RuntimeError: svd_cuda: For batch 0: U(65,65) is zero, singular U.`

This unexpected (and misleading) error message comes from a special `batchCheckErrors()` info-tensor function overload that does not check for the "svd" in the name string, see https://github.com/pytorch/pytorch/blob/ac67cda272b0856c1be3ebfebc324e0b3bab286f/aten/src/ATen/native/LinearAlgebraUtils.h#L242-L254
If you look above and below e.g. `batchCheckErrors(std::vector<int64_t>& infos, ...)` and `void singleCheckErrors()`  both have the case:
```
if (strstr(name, "svd")) {
     AT_ERROR(name, ": the updating process of SBDSDC did not converge (error: ", info, ")");
}
```


I guess the tensor-info `batchCheckErrors()` overload it is called from `_svd_helper_cuda_lib()`: https://github.com/pytorch/pytorch/blob/3f0b081636cd2d84c502f62b38b541ea961ff447/aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu#L456-L490

**Comment by IvanYashchuk** [comment]
Thank you, @andreaskoepf, for reporting the issue of incorrect error messages! We will have it fixed in the future PyTorch release. Unfortunately, the bugfixes would not be backported to older versions.

**Comment by lhyfst** [comment]
> I am using this, since it is not solved
> 
> ```
>     try:
>         u, s, v = torch.svd(L)
>     except:                     # torch.svd may have convergence issues for GPU and CPU.
>         u, s, v = torch.svd(L + 1e-4*L.mean()*torch.rand(l, h))
> ```

What is h in `torch.rand(l, h)`?

**Comment by andreaskoepf** [comment]
> What is h in `torch.rand(l, h)`?

To add some noise you could use `torch.rand_like(L)`.

For support & coding questions please refer to the [PyTorch Forums](https://discuss.pytorch.org/) & see the docs, e.g. [torch.rand](https://pytorch.org/docs/stable/generated/torch.rand.html?highlight=rand#torch.rand).

**Comment by lhyfst** [comment]
> > What is h in `torch.rand(l, h)`?
> 
> To add some noise you could use `torch.rand_like(L)`.
> 
> For support & coding questions please refer to the [PyTorch Forums](https://discuss.pytorch.org/) & see the docs, e.g. [torch.rand](https://pytorch.org/docs/stable/generated/torch.rand.html?highlight=rand#torch.rand).

Thank you

**Comment by a-r-j** [comment]
I had some luck with:


```python
from tenacity import retry, stop_after_attempt


@retry(stop=stop_after_attempt(32)) # Or some other value
def func_with_svd(L: torch.Tensor):
    try:
        u, s, v = torch.svd(L)
    except:                     # torch.svd may have convergence issues for GPU and CPU.
        u, s, v = torch.svd(L + 1e-4*L.mean()*torch.rand(l, h))
    ...
```
